### DISCUSSION TITLE
How to define route parameters?

### DISCUSSION THREAD

**Discussion by vedun** [RobinTail/express-zod-api#73] [Q&A] [answered]
Hello.
I have a question, it is possible to configure route parameter in endpoint `url`? For example for route `/user/:id` i want to get `id` parameter somewhere in handler parameter, like it is happens for query parameters.
Thanks.

**Comment by RobinTail** [comment, marked as answer]
Hello @vedun ,

Thank you for the question.
The path is being constructed by concatenation of properties defined in `Routing` object.
So you just need to have an `:id` property inside of `user`. In order to do that you need to wrap it in quotes.

To obtain the value of URL param you will also require a middleware that will provide your endpoint handler with such option.

Here is a complete example for you:

```typescript
const paramsProviderMiddleware = createMiddleware({
  input: z.object({}),
  middleware: async ({request}) => ({
    params: request.params
  })
});

const endpointsFactory = new EndpointsFactory();

const myEndpoint = endpointsFactory
  .addMiddleware(paramsProviderMiddleware)
  .build({
    methods: ['get'],
    input: z.object({}),
    output: z.object({}),
    handler: async ({options, logger}) => {
      logger.debug('id', options.params.id); // the value of :id in the path
      return {};
    }
  });

const myRouting: Routing = {
  user: {
    ':id': myEndpoint // <— /user/:id 
  }
};

createServer(config, myRouting);
```

Please let me know how it works for you.

**Comment by vedun** [comment]
@RobinTail  thank you for your answer. Probably it is good solution for me.

**Comment by RobinTail** [comment]
@vedun , starting version 3.1.x you no longer need `paramsProviderMiddleware`.

**Reply by vedun** [reply]
I will answer when come back on this task.
